pkg/ipam-node/store/store.go

```diff
@@ -0,0 +1,315 @@
+	m.lock.Lock()
+	logger := logr.FromContextOrDiscard(ctx)
+	if m.persistentData == nil {
+		diskData, err := m.loadFromDisk()
+		if err != nil {
+			m.lock.Unlock()
+			logger.Error(err, "failed to load store data from the disk")
+			return nil, err
+		}
+		m.persistentData = diskData
+	}
+	return newStore(m.storeFile, m.persistentData, m.lock, logger), nil
+}
+
+func (m *manager) loadFromDisk() (*types.Root, error) {
+	dst := &types.Root{}
+	data, err := os.ReadFile(m.storeFile)
+	if err != nil && !errors.Is(err, os.ErrNotExist) {
+		return nil, fmt.Errorf("failed to read store persistedData: %v", err)
+	}
+	if len(data) != 0 {
+		if err := json.Unmarshal(data, dst); err != nil {
+			return nil, fmt.Errorf("failed to unmarshal store persistedData: %v", err)
+		}
+		if dst.Checksum == 0 {
+			return nil, fmt.Errorf("checksum not set in store file")
+		}
+		tmpChecksum := dst.Checksum
+		dst.Checksum = 0
+		if err := tmpChecksum.Verify(dst); err != nil {
+			return nil, fmt.Errorf("store file corrupted, checksum mismatch")
+		}
+		dst.Checksum = tmpChecksum
+	} else {
+		dst = types.NewRoot()
+	}
+	if dst.Pools == nil {
+		dst.Pools = map[string]types.PoolReservations{}
+	}
+	return dst, nil
+}
+
+func newStore(storeFile string, persistentData *types.Root, lock *sync.Mutex, logger logr.Logger) *store {
+	s := &store{
+		log:           logger,
+		lock:          lock,
+		storeFile:     storeFile,
+		persistedData: persistentData,
+		tmpData:       persistentData.DeepCopy(),
+	}
+	return s
+}
+
+// default store implementation
+type store struct {
+	log       logr.Logger
+	lock      *sync.Mutex
+	storeFile string
+	// manager holds this data instance
+	persistedData *types.Root
+	// temp data, this data is local for store instance
+	tmpData    *types.Root
+	isModified bool
+	isClosed   bool
+}
+
+// Commit is the Store interface implementation for store
+func (s *store) Commit() error {
+	defer s.markClosed()
+	defer s.lock.Unlock()
+	if !s.isModified {
+		return nil
+	}
+	if err := s.writeToDisk(s.tmpData); err != nil {
+		s.log.Error(err, "failed to write data to disk")
+		return err
+	}
+	s.persistTMP()
+	return nil
+}
+
+func (s *store) persistTMP() {
+	*s.persistedData = *s.tmpData.DeepCopy()
+}
+
+// Cancel is the Store interface implementation for store
+func (s *store) Cancel() {
+	defer s.markClosed()
+	defer s.lock.Unlock()
+}
+
+func (s *store) markClosed() {
+	s.isClosed = true
+}
+
+func (s *store) checkClosed() {
+	if s.isClosed {
+		panic("operation was performed on the closed store")
+	}
+}
+
+// Reserve is the Store interface implementation for store
+func (s *store) Reserve(pool string, id string, ifName string, meta types.ReservationMetadata, address net.IP) error {
+	s.checkClosed()
+	reservationKey := s.getKey(id, ifName)
+	poolData := s.getPoolData(pool, s.tmpData)
+	_, exist := poolData.Entries[reservationKey]
+	if exist {
+		return ErrReservationAlreadyExist
+	}
+	duplicateIP := false
+	for _, r := range poolData.Entries {
```

Comment:
to make this more efficient, we can have an in-memory set of IPs per pool to check for duplicates against it.

not sure its really an issue at this point.

Reply:
I think no need for now